Sources/AblyChat/RoomLifecycleManager.swift

```diff
@@ -0,0 +1,297 @@
+    var errorReason: ARTErrorInfo? { get async }
+}
+
+internal actor RoomLifecycleManager<Channel: RoomLifecycleContributorChannel> {
+    /// A realtime channel that contributes to the room lifecycle.
+    internal struct Contributor {
+        /// The room feature that this contributor corresponds to. Used only for choosing which error to throw when a contributor operation fails.
+        internal var feature: RoomFeature
+
+        internal var channel: Channel
+    }
+
+    internal private(set) var current: RoomLifecycle
+    internal private(set) var error: ARTErrorInfo?
+
+    private let logger: InternalLogger
+    private let clock: SimpleClock
+    private let contributors: [Contributor]
+
+    internal init(
+        contributors: [Contributor],
+        logger: InternalLogger,
+        clock: SimpleClock
+    ) {
+        self.init(
+            current: nil,
+            contributors: contributors,
+            logger: logger,
+            clock: clock
+        )
+    }
+
+    #if DEBUG
+        internal init(
+            testsOnly_current current: RoomLifecycle? = nil,
+            contributors: [Contributor],
+            logger: InternalLogger,
+            clock: SimpleClock
+        ) {
+            self.init(
+                current: current,
+                contributors: contributors,
+                logger: logger,
+                clock: clock
+            )
+        }
+    #endif
+
+    private init(
+        current: RoomLifecycle?,
+        contributors: [Contributor],
+        logger: InternalLogger,
+        clock: SimpleClock
+    ) {
+        self.current = current ?? .initialized
+        self.contributors = contributors
+        self.logger = logger
+        self.clock = clock
+    }
+
+    // TODO: clean up old subscriptions (https://github.com/ably-labs/ably-chat-swift/issues/36)
+    private var subscriptions: [Subscription<RoomStatusChange>] = []
+
+    internal func onChange(bufferingPolicy: BufferingPolicy) -> Subscription<RoomStatusChange> {
+        let subscription: Subscription<RoomStatusChange> = .init(bufferingPolicy: bufferingPolicy)
+        subscriptions.append(subscription)
+        return subscription
+    }
+
+    /// Updates ``current`` and ``error`` and emits a status change event.
+    private func changeStatus(to new: RoomLifecycle, error: ARTErrorInfo? = nil) {
+        logger.log(message: "Transitioning from \(current) to \(new), error \(String(describing: error))", level: .info)
+        let previous = current
+        current = new
+        self.error = error
+        let statusChange = RoomStatusChange(current: current, previous: previous, error: error)
+        emitStatusChange(statusChange)
+    }
+
+    private func emitStatusChange(_ change: RoomStatusChange) {
+        for subscription in subscriptions {
+            subscription.emit(change)
+        }
+    }
+
+    /// Implements CHA-RL1’s `ATTACH` operation.
+    internal func performAttachOperation() async throws {
```

Comment:
**suggestion:** Consider breaking down complex methods into smaller, focused functions

The `performAttachOperation` and `performDetachOperation` methods are quite long and handle multiple cases. Breaking these down into smaller, more focused methods could improve readability and maintainability.